cpp/src/io/orc/orc.cpp

```diff
@@ -209,43 +209,55 @@ void ProtobufWriter::put_row_index_entry(int32_t present_blk,
                                          int32_t data_ofs,
                                          int32_t data2_blk,
                                          int32_t data2_ofs,
-                                         TypeKind kind)
+                                         TypeKind kind,
+                                         ColStatsBlob const* stats)
 {
   size_t sz = 0, lpos;
-  putb(1 * 8 + PB_TYPE_FIXEDLEN);  // 1:RowIndex.entry
+  put_byte(1 * 8 + PB_TYPE_FIXEDLEN);  // 1:RowIndex.entry
   lpos = m_buf->size();
-  putb(0xcd);                      // sz+2
-  putb(1 * 8 + PB_TYPE_FIXEDLEN);  // 1:positions[packed=true]
-  putb(0xcd);                      // sz
+  put_byte(0xcd);                      // sz+2
+  put_byte(1 * 8 + PB_TYPE_FIXEDLEN);  // 1:positions[packed=true]
+  put_byte(0xcd);                      // sz
   if (present_blk >= 0) sz += put_uint(present_blk);
   if (present_ofs >= 0) {
     sz += put_uint(present_ofs) + 2;
-    putb(0);  // run pos = 0
-    putb(0);  // bit pos = 0
+    put_byte(0);  // run pos = 0
+    put_byte(0);  // bit pos = 0
   }
   if (data_blk >= 0) { sz += put_uint(data_blk); }
   if (data_ofs >= 0) {
     sz += put_uint(data_ofs);
     if (kind != STRING && kind != FLOAT && kind != DOUBLE && kind != DECIMAL) {
-      putb(0);  // RLE run pos always zero (assumes RLE aligned with row index boundaries)
+      // RLE run pos always zero (assumes RLE aligned with row index boundaries)
+      put_byte(0);
       sz++;
       if (kind == BOOLEAN) {
-        putb(0);  // bit position in byte, always zero
+        // bit position in byte, always zero
+        put_byte(0);
         sz++;
       }
     }
   }
-  if (kind !=
-      INT)  // INT kind can be passed in to bypass 2nd stream index (dictionary length streams)
-  {
+  // INT kind can be passed in to bypass 2nd stream index (dictionary length streams)
+  if (kind != INT) {
     if (data2_blk >= 0) { sz += put_uint(data2_blk); }
     if (data2_ofs >= 0) {
       sz += put_uint(data2_ofs) + 1;
-      putb(0);  // RLE run pos always zero (assumes RLE aligned with row index boundaries)
+      // RLE run pos always zero (assumes RLE aligned with row index boundaries)
+      put_byte(0);
     }
   }
-  m_buf->data()[lpos]     = (uint8_t)(sz + 2);
+  // size of the field 1
   m_buf->data()[lpos + 2] = (uint8_t)(sz);
+
+  if (stats != nullptr) {
+    sz += put_uint(2 * 8 + PB_TYPE_FIXEDLEN);
```

Comment:
2 and 8 seem magical here. Should they be named with a constexpr to add some clarity to this calculation?

Maybe this section needs a comment like
```
// data stream is a 4 byte type followed by a 4 byte size followed by the actual stat data block
```

Reply:
2 is a field number, from the ORC protobuf spec. I think the meaning is a lot clearer now, with the use of `encode_field_number`.

Reply:
This is much nicer, thanks!